File: bfs/sliding_puzzle/sliding_puzzle.py

def _swap_chars(string: str, idx_1: int, idx_2: int) -> str:
    chars = list(string)
    chars[idx_1], chars[idx_2] = chars[idx_2], chars[idx_1]
    return "".join(chars)

def sliding_puzzle(board: list[list[int]]) -> int:  # LeetCode Q.773.
    init_state = ""  # Vectorize board w.r.t. rows into string.
    for row_idx in range(2):
        for col_idx in range(3):
            init_state += str(board[row_idx][col_idx])

    visited_states: set[str] = set()  # Format: vectorize board w.r.t. rows into string.

    # Format: (current state, moves). Start from the solved state.
    queue: list[tuple[str, int]] = [("123450", 0)]
    while queue:
        state, moves = queue.pop(0)
        if state == init_state:
            return moves

        moves += 1
        visited_states.add(state)

        for idx in range(6):
            if state[idx] == "0":  # Locate the idx containing "0".
                if idx % 3 != 2:
                    state = _swap_chars(state, idx, idx + 1)
                    if state not in visited_states:
                        queue.append((state, moves))

                    state = _swap_chars(state, idx, idx + 1)  # Swap back for later usage.

                if idx % 3 != 0:
                    state = _swap_chars(state, idx, idx - 1)
                    if state not in visited_states:
                        queue.append((state, moves))

                    state = _swap_chars(state, idx, idx - 1)  # Swap back for later usage.

                if idx < 3:
                    state = _swap_chars(state, idx, idx + 3)
                    if state not in visited_states:
                        queue.append((state, moves))

                    state = _swap_chars(state, idx, idx + 3)  # Swap back for later usage.

                if idx >= 3:
                    state = _swap_chars(state, idx, idx - 3)
                    if state not in visited_states:
                        queue.append((state, moves))

                break

    return -1  # Can't be solved.

File: bfs/sliding_puzzle/sliding_puzzle.cpp
#include <vector>
#include <string>
#include <unordered_set>
#include <queue>
using namespace std;

int sliding_puzzle(vector<vector<int>> &board)
{
    // LeetCode Q.773.
    string init_state = ""; // Vectorize board w.r.t. rows into string.
    for (int row_idx = 0; row_idx < 2; row_idx++)
        for (int col_idx = 0; col_idx < 3; col_idx++)
            init_state += to_string(board[row_idx][col_idx]);

    // Format: vectorize board w.r.t. rows into string.
    unordered_set<string> visited_states;

    queue<pair<string, int>> queue; // Format: {current state, moves}.
    queue.push({"123450", 0});      // Start from the solved state.
    while (!queue.empty())
    {
        auto [state, moves] = queue.front();
        if (state == init_state)
            return moves;

        moves++;
        visited_states.insert(state);
        queue.pop();

        for (int idx = 0; idx < 6; idx++)
            if (state[idx] == '0')
            { // Locate the idx containing '0'.
                if (idx % 3 != 2)
                {
                    swap(state[idx], state[idx + 1]);
                    if (visited_states.find(state) == visited_states.end())
                        queue.push({state, moves});

                    swap(state[idx], state[idx + 1]); // Swap back.
                }

                if (idx % 3 != 0)
                {
                    swap(state[idx], state[idx - 1]);
                    if (visited_states.find(state) == visited_states.end())
                        queue.push({state, moves});

                    swap(state[idx], state[idx - 1]); // Swap back.
                }

                if (idx < 3)
                {
                    swap(state[idx], state[idx + 3]);
                    if (visited_states.find(state) == visited_states.end())
                        queue.push({state, moves});

                    swap(state[idx], state[idx + 3]); // Swap back.
                }

                if (idx >= 3)
                {
                    swap(state[idx], state[idx - 3]);
                    if (visited_states.find(state) == visited_states.end())
                        queue.push({state, moves});

                    swap(state[idx], state[idx - 3]); // Swap back.
                }
                break;
            }
    }

    return -1; // Can't be solved.
}
